### PR TITLE
New wasm image and WasmPlugin custom resource support

### PR DESCRIPTION
**Pull request by azgabur**
This PR is dependent on yet unreleased wasm image. 
(for testing purposes, @pehala 's image can be used `quay.io/phala/threescale-wasm-auth:v16`)

- Adds support for new WasmPlugin
- Fixes gateways resource not deleting
- Removes mapping rules synchronization, as its not needed in new wasm image 
- Removes backend token synchronization, as its also not needed
- Remove special status code handling in tests as now correct status codes are returned
- Reduce openshift calls in credentials synchronization and rollout only if they were changed
- Add image pull secret support for wasm (_currently untested_)

### CODE REVIEW

File: testsuite/gateways/wasm/__init__.py [hunk 1]
```diff
@@ -96,11 +96,10 @@ def remove_policy(self, name: str):
     # pylint: disable=unused-argument, too-many-arguments
     def _create_api_client(self, application, endpoint, verify, cert=None, disable_retry_status_list=None):
         ext = self.extensions[application.service["id"]]
-        ext.synchronise_mapping_rules()
-        ext.synchronise_credentials()
+        if ext.synchronise_credentials():
+            ext.httpbin.deployment(f"dc/{ext.httpbin_name}").rollout()
+            ext.httpbin.deployment(f"dc/{ext.httpbin_name}").wait_for()
```
Comment (pehala) [inline testsuite/gateways/wasm/__init__.py:101]: I believe rollout also waits for the resource

File: testsuite/gateways/wasm/extension.py [hunk 2]
```diff
@@ -106,14 +104,27 @@ def synchronise_mapping_rules(self):
...
+    def replace_credentials(self, credentials):
+        """Replaces previous credentials and deploys new into the extension.
+            `credentials` contains list of credentials with attributes:
             label: one of "user_key" or "app_id" or "app_key",
-            credential_location: one of "query_string", "header" or "authorization"
+            credential_location: one of "query_string", "header", "authorization", or "oidc"
             key: search key
         """
         ops = []
+        for label in ["user_key", "app_id", "app_key"]:
```
Comment (pehala) [inline testsuite/gateways/wasm/extension.py:115]: [Minor] Use set

File: testsuite/resources/service_mesh/plugin.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+objects:
+- apiVersion: extensions.istio.io/v1alpha1
+  kind: WasmPlugin
+  metadata:
+    name: ${NAME}
+    labels:
+      app: ${LABEL}
+  spec:
+    url: oci://${IMAGE}
+    phase: AUTHZ
```
Comment (pehala) [inline testsuite/resources/service_mesh/plugin.yaml:14]: If used with the released image, we will also need to add a field containing a pullSecret.

File: testsuite/gateways/wasm/__init__.py (hunk 1 above)
Comment (azgabur) [reply testsuite/gateways/wasm/__init__.py]: If so, I removed the wait

File: testsuite/gateways/wasm/extension.py (hunk 2 above)
Comment (azgabur) [reply testsuite/gateways/wasm/extension.py]: Addressed

Comment (azgabur) [reply testsuite/gateways/wasm/extension.py]: It doesnt really make a difference and set usage generates lint warning 
` C0208: Use a sequence type when iterating over values (use-sequence-for-iteration)`
Reverting

File: testsuite/resources/service_mesh/plugin.yaml (hunk 3 above)
Comment (azgabur) [reply testsuite/resources/service_mesh/plugin.yaml:15]: Added support for this

**Review by pehala — APPROVED** [review]

File: config/settings.yaml [hunk 4]
```diff
@@ -13,6 +13,7 @@ default:
         template: apicast.yml
       WASMGateway:
         image: "registry.redhat.io/openshift-service-mesh/3scale-auth-wasm-rhel8:0.0.1"
+        pull_secret: "pull-secret"  # name of pull secret resource in same namespace as httpbin
```
Comment (dhlavac) [inline config/settings.yaml:16]: Please add also WASMGate configuration to settings tlp.

Comment (pehala) [reply config/settings.yaml:16]: I agree, however I would add it in a separate PR

Comment (azgabur) [reply config/settings.yaml:16]: Opened new PR #461 for this

**Review by dhlavac — APPROVED** [review]

**Pull request merged**